CLDSRV-636: SSE with both internal & external KMS (HF 9.2.0.12)

BourgoisMickael opened the pull request:
This is generic but simplified for file to aws use case and will be extended in latest version to handle any backend.

- Store all KMS keys with an arn describing backend provider (see https://github.com/scality/Arsenal/pull/2349)
  - option (`hideScalityArn`) to keep backward compatibility
 - Migrate KMS Keys with `sseMigration` config
   - For bucket, replace `file` key with `aws` key
   - For object reformat key into an arn
- Allow usage of keys from previous KMS provider (`file`) to decript and read files (or for ongoing MPU)


### Overall example of a new Key Format and migration

```mermaid
flowchart TD
    GetObject:::red --> bktArn
    objArn --NO GetObject--> objUpdate[Update Object: Prefix KMS key with arn]:::red
    %% Get 0-1 Links

    %% Common route
    bktArn{Bucket SSE is arn?}:::purple
        --NO--> bktKey["
            Create new external KMS Key with arn prefix
            Update Bucket attribute
        "]:::purple
    bktArn
        --Then--> objArn{Object SSE is arn?}:::purple
        --Then--> ...:::purple
        --> KMS
    subgraph KMS[KMS#nbsp;Object#nbsp;Decrypt/Encrypt]
        extract[Extract KMS provider and KeyId from arn]:::purple
        --> client[Pick or instantiate KMS client based on provider from Key arn]:::purple
        --> send["
            Backend Decrypts/Encrypts with KeyId
            But if Arn is sent, client still extract KeyId from arn
        "]:::purple
    end
    %% 2-7 Links common

    PutObject:::blue --> bktArn
    objArn --NO PutObject--> PutPrefix[Prefix KMS Key with arn if missing]:::blue
    %% PUT 8-9 Links

    classDef red stroke:red
    classDef blue stroke:blue
    classDef purple stroke:purple

    linkStyle 0,1 stroke:red
    linkStyle 8,9 stroke:blue
    linkStyle 2,3,4,5,6,7 stroke:purple
```

BourgoisMickael committed:
CLDSRV-625: No scality-kms warning if not conf

Avoid print the warning if scality-kms is not configured:

```json
{"name":"S3","time":1742220545805,"error":{"code":"MODULE_NOT_FOUND","requireStack":["/home/mickael/scality/cloudserver/lib/kms/wrapper.js","/home/mickael/scality/cloudserver/lib/data/wrapper.js","/home/mickael/scality/cloudserver/lib/utilities/healthcheckHandler.js","/home/mickael/scality/cloudserver/lib/utilities/internalHandlers.js","/home/mickael/scality/cloudserver/lib/server.js","/home/mickael/scality/cloudserver/index.js"]},"level":"warn","message":"scality kms unavailable. Using file kms backend unless mem specified.","hostname":"mickael-xps","pid":139701}
```

(cherry picked from commit 9e8c505dd4a7dcc72f83fc809801e72aea9eb650)

BourgoisMickael committed:
CLDSRV-635: Fix ubuntu 20 in CI

> This is a scheduled Ubuntu 20.04 brownout. Ubuntu 20.04 LTS runner will be removed on 2025-04-15. For more details, see https://github.com/actions/runner-images/issues/11101

(cherry picked from commit 5c4a1b37ac01352c85fd4420db506dbbad3c5728)

BourgoisMickael committed:
CLDSRV-636: Prefix KMS Key with an arn

BourgoisMickael committed:
CLDSRV-636: Option for KMS to hide scality arn

To avoid breaking changes for clients

BourgoisMickael committed:
CLDSRV-636: SSE Migration function

BourgoisMickael committed:
CLDSRV-636: Migrate SSE on object(Get|Head|Put)

BourgoisMickael committed:
CLDSRV-636: Handle multi KMS clients for migration

BourgoisMickael committed:
CLDSRV-636: Bump arsenal

bert-e commented:
# Hello bourgoismickael,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-636 contains:


* *None*



Considering where you are trying to merge, I expected to find at least:



* `7.70.21.1`


Please check the `Fix Version/s` of CLDSRV-636, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

nicolas2bert commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,140 @@
+
+    // Note: if migration is from an ext to an ext, absence or arn is not enough
+    // a comparison of arn will be necessary
+    const updateMaster = masterKey && !masterKey.startsWith('arn:scality:kms:');
```
isScalityKmsArn(...) from arsenal could be used here?

nicolas2bert commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,140 @@
+    const masterKey = sse.masterKeyId;
+    const configuredKey = sse.configuredMasterKeyId;
+
+    // Note: if migration is from an ext to an ext, absence or arn is not enough
```
absence **of** arn prefix?

nicolas2bert commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,140 @@
+    // this should trigger vault account key update as well
+    return kms.createBucketKey(bucket, log, (err, { masterKeyArn }) => {
+        if (err) {
+            return cb(err, bucket);
```
should we log something?

BourgoisMickael replied:
errors should already be logged by kms implementation. But we could append a message to the error to describe the error is due to the SSE migration

nicolas2bert commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,140 @@
+            sse.masterKeyId = masterKeyArn;
+        }
+        if (updateConfigured) {
+            sse.configuredMasterKeyId = masterKeyArn;
```
Are `sse.masterKeyId` and `sse.configuredMasterKeyId` not suppose to be different?
Does it break something if we have them both be the same value?

BourgoisMickael replied:
Yes they should be different, I need to fix this

BourgoisMickael replied:
Actually nothing should break, right now they are different because they are set at different times, but having the same key is fine

nicolas2bert commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,140 @@
+const kms = require('../../../kms/wrapper');
+const metadata = require('../../../metadata/wrapper');
+
+// Bucket need a key from the new KMS, not a simple reformating
```
If we migrate from an external KMS to the same external KMS, we don't need to recreate master keys. Right? 
We don't handle the case here because it is for Outscale only that does an internal to external migration?

At the end the issue is not breaking since we will just generate duplicated KMS master key, right?

BourgoisMickael replied:
For outscale it's internal to external, but there should be validation in place to prevent configuring a migration to the same provider.

BourgoisMickael committed:
CLDSRV-636: Speed a test up with fake timer (10s)

BourgoisMickael committed:
CLDSRV-636: Fix unit tests with KMS arn prefix

nicolas2bert commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,140 @@
+
+// Only reformat the key, don't generate a new one.
+// Use opts.skipObjectUpdate to only prepare objMD without sending the update to metadata
+// if an putObject is expected later in call flow. (Downside: update skipped if error)
```
Do you mean: "if a put object metadata (i.e. `metadata.putObjectMD`)"

BourgoisMickael replied:
Yes

nicolas2bert commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,140 @@
+    // Doesn't seem to be used but update as well
+    for (const dataLocator of objMD.location || []) {
+        if (dataLocator.masterKeyId) {
+            dataLocator.masterKeyId = `${keyArnPrefix}${dataLocator.masterKeyId}`;
```
Could it be that the key and dataLocator.masterKeyId were, or could have been, updated separately?

- What if key has the old format, and dataLocator.masterKeyId has the new one?
In that case, dataLocator.masterKeyId might be updated with the same prefix multiple times.

- What if key has the new format, and dataLocator.masterKeyId still uses the old one?
In that case, dataLocator.masterKeyId might never get updated to the new format.

BourgoisMickael replied:
I don't think so, they are not used to get the object as the GetObject rewrite those to decrypt the data: https://github.com/scality/cloudserver/blob/0f788abd327750d2a58d630babf1df3cfb423bca/lib/api/objectGet.js#L247-L251

I'll check why they are set on objects sometimes

BourgoisMickael replied:
For any GET | Website | Copy the dataLocator sse is always rewritten with the metadata sse.

I suspect it should not be saved into DB but it's set on PUT for the dataWrapper to encrypt the data but in some cases end up being written into metadata

nicolas2bert commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,140 @@
+    objMD.originOp = 's3:ObjectCreated:Copy';
+    const params = getVersionSpecificMetadataOptions(objMD, config.nullVersionCompatMode);
+
+    // TODO: what about CRR ?
```
What case should be handled for CRR?

BourgoisMickael replied:
make sure this internal migration is ignored for CRR and bucket notification.

But outscale don't use those, so it could be done later

nicolas2bert commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    const keyArnPrefix =
+        `arn:scality:kms:${previousKeyType}:${previousKeyProtocol}:${previousKeyProvider}:key/`;
```
Should we centralize the creation of the prefix in one place? I saw it in different classes/repos.

nicolas2bert commented on lib/api/objectGet.js:
```diff
@@ -51,6 +52,7 @@ function objectGet(authInfo, request, returnTagCount, log, callback) {
     };
 
     return metadataValidateBucketAndObj(mdValParams, log,
+    (err, bucket, objMD) => updateEncryption(err, bucket, objMD, objectKey, log, {},
```
Should we fail the GET/HEAD object call if update encryption fails?

BourgoisMickael replied:
Right now it will fail, the error is passed to the next callback.

But we could provide a flag through config to ignore migration failure and still return the object.

The only reason where the update could fail but not the get is if you can't write, because disk is full but still wants to read from it.

nicolas2bert commented on lib/kms/wrapper.js:
```diff
@@ -309,17 +377,21 @@ class KMS {
         };
+
+        // shadowing global client for key - implName already used can't be shadowed here
+        const { client, implName: _impl, key } = getClientForKey(serverSideEncryptionInfo.masterKeyId);
```
We always need to check the client because there is no way to know that a migration never happened before? If we had a way to know that no migration happened before, we would be safe to use the default KMS, right? How is the impact of getClientForKey on the GET object latency?

BourgoisMickael replied:
If we know the migration never happened before, we'd rather know we should use the `previous` KMS instead of the new default external KMS.

But the way to know for now for outscale is if the scality kms arnPrefix is present or not.

I've also pre instantiated the previous KMS as `file` is the only allowed for now in migration, to avoid latency of first GET after startup.

The impact on GET object latency is:
- extract key detail from arn: `split('/')` and `split(':')` on the key: https://github.com/scality/Arsenal/pull/2349/files#diff-1c24d37a19bc08405b548ef9b5ac12a2ef83b7edb3a2ee62cc8426c7df8e5b95R122-R137
- if key is scality arnPrefix
  - validate the key parts: https://github.com/scality/Arsenal/pull/2349/files#diff-1c24d37a19bc08405b548ef9b5ac12a2ef83b7edb3a2ee62cc8426c7df8e5b95R145-R167
  - concatenate the `clientIdentifier` string: https://github.com/scality/cloudserver/pull/5788/files#diff-c506ce0254987c99cd5ff5a10cd1ecabc6e1bc59e643fe2401103b9feb7b27a6R131
  - return the client and KeyId
- if key is not scality arnPrefix
  - return the client (default or previous based on `sseMigration` configuration presence) and the KeyId

But as the GetObject migrate to new scality arn format, it should be expected the first Get on a bucket and object is slower for migration, and then we fall in the case where it's a scality arnPrefix

nicolas2bert commented on lib/kms/wrapper.js:
```diff
@@ -196,10 +261,13 @@ class KMS {
         }
+        // shadowing global client for key
+        // but should not happen to cipher for another client as Puts should use current KMS
+        const { client, implName, key } = getClientForKey(masterKeyId);
```
Is getClientForKey needed for PUT object? We know that the client will always be the default one. What is the impact on our PUT object latency?

BourgoisMickael replied:
It's needed if kms key is passed in scality arn format on a PUT request. Then the function will extract the KeyId from the arn and validate the arn scality format.

It also returns the client identifier so we can block usage of other client like if an arn for the `file` KMS is used when there is a configured external KMS, it should be forbidden: https://github.com/scality/cloudserver/pull/5788/files#diff-c506ce0254987c99cd5ff5a10cd1ecabc6e1bc59e643fe2401103b9feb7b27a6R332-R339

___

And for any future case where multiple KMS providers can be used it will become useful

BourgoisMickael replied:
As for PUT latency, it will be same as GET:

The impact on PUT object latency is:

- extract key detail from arn: split('/') and split(':') on the key: https://github.com/scality/Arsenal/pull/2349/files#diff-1c24d37a19bc08405b548ef9b5ac12a2ef83b7edb3a2ee62cc8426c7df8e5b95R122-R137
- if key is scality arnPrefix
  - validate the key parts: https://github.com/scality/Arsenal/pull/2349/files#diff-1c24d37a19bc08405b548ef9b5ac12a2ef83b7edb3a2ee62cc8426c7df8e5b95R145-R167
  - concatenate the clientIdentifier string: https://github.com/scality/cloudserver/pull/5788/files#diff-c506ce0254987c99cd5ff5a10cd1ecabc6e1bc59e643fe2401103b9feb7b27a6R131
  - return the client and KeyId
- if key is not scality arnPrefix (should not happen as the PUT will prefix the key with a scality arn)
  - return the client (default or previous based on sseMigration configuration presence) and the KeyId

nicolas2bert commented on lib/kms/wrapper.js:
```diff
@@ -10,49 +10,106 @@ const KMIPClient = require('arsenal').network.kmipClient;
+
+// TODO store KMS identifier differently to avoid concatenations + split of arn
+// multiple times for a request
+function getClientForKey(key) {
```
Can we be certain that key always follows the new format?

BourgoisMickael replied:
We should be, but I need to validate the format of the key passed to  PutObject if a scality arn is provided. The other places it's generated internally so we are good

BourgoisMickael replied:
I've added a validation function if the key matches the scality arn format.

Now we are sure the format is valid before using it with the KMS

BourgoisMickael committed:
CLDSRV-636: Fix functional tests encryption

BourgoisMickael committed:
CLDSRV-636: Fix SSE with MPU & CopyObject headers

Issue: S3C-9996

MPU was not accepting configuredMasterKeyId (from object or bucket)
for Create and Complete MPU.

But used the configuredMasterKeyId (from object or bucket)
for UploadPart making it possible to have parts encrypted with a
different key than the key in object metadata after completion.

Resulting in potential error on GetObject later:
 - If bucket had configuredMasterKeyId but no masterKeyId
   (no AES before or aws:kms without configuredKey)
   it would crash on decryption.
 - If bucket had configuredMasterKeyId and a masterKeyId
   it would be used to decrypt parts potentially
   encrypted with a configuredMasterKeyId, decrypting gibberish
 - On UploadPart object level encryption could be provided (outside sdk/cli)
   which is bad.

___

The expected behavior fixed here is to use the provided SSE on CreateMPU
and then for UploadPart and CompleteMPU use the SSE provided at CreateMPU.

The MPU calls also return the SSE headers now.
The CopyObject returned SSE headers are fixed as well, the configuredMasterKeyId was ignored

This fix allow seamless continuation of ongoing MPU during SSE migration.

___

There is no migration to fix already existing completed MPU.

BourgoisMickael committed:
CLDSRV-636: Bump version

anurag4DSB commented on lib/Config.js:
```diff
@@ -1485,6 +1517,61 @@ class Config extends EventEmitter {
+            this.sseMigration.previousKeyType = previousKeyType;
+
+            let expectedProtocol;
+            if (previousKeyType === KmsType.int) {
```
We should use `internal` here as `int` can give the impression of integers.
Other places in the config as well.

BourgoisMickael replied:
It's been shipped like that, but each values will be documented as they don't represent the regular meaning in an aws arn. 
And it should not be customer facing, by default the arn is stripped when the key is returned to the client

anurag4DSB commented on lib/kms/wrapper.js:
```diff
@@ -10,49 +10,157 @@ const KMIPClient = require('arsenal').network.kmipClient;
+    try {
+        // eslint-disable-next-line import/no-unresolved
+        const ScalityKMS = require('scality-kms');
+        // TODO add scality arn in scality-kms
```
Do we still need this todo?

BourgoisMickael replied:
This is not needed for outscale but this is a reminder to be done in 9.5 when cherry-picking.
Here ticket: https://scality.atlassian.net/browse/KMS-7

BourgoisMickael replied:
Dropped the comment: https://github.com/scality/cloudserver/pull/5788/commits/fd1ca6bd2ef29c59a6d4c478c822ffd818aa7e0e

anurag4DSB commented on lib/Config.js:
```diff
@@ -1509,6 +1517,61 @@ class Config extends EventEmitter {
+            const { previousKeyType, previousKeyProtocol, previousKeyProvider } = config.sseMigration;
+            if (!previousKeyType) {
+                assert.fail(
+                    'NotImplemented: No dynamic KMS key migration. Set sseMigration.previousKeyType');
```
NIT: Will it be helpful to use AWS (arsenal) Not implemented error with a customized description for this?
I know we do not have to comply with AWS specs here, that's why its a nit.

BourgoisMickael replied:
No need here, I want to fail the startup of cloudserver, like any other assert on config.

But for 9.5 it will be validated in Federation beforehand

anurag4DSB commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,144 @@
+            sse.configuredMasterKeyId = masterKeyArn;
+        }
+
+        log.info('updating bucket encryption', { oldKey: masterKey || configuredKey, newKey: masterKeyArn });
```
This might be worth a debug log and not info log.
For info log we try to make sure we only have entry and exit logs.
I am curious why info will be helpful here for the extra log it might add.

BourgoisMickael replied:
Usually logger is in info level and I wanted to be sure the print the log as this is something that should be happen once per bucket and once per object (below), so we can clearly see the migration is happening in logs without having to enable debug logs.

BourgoisMickael replied:
This is actually useful to torubleshoot the migration specially on special hotfix versions.

We should rather keep them as info for a while with our first clients using this. Maybe later it could me moved to debug

anurag4DSB replied:
Okay, in general we should enable debug logs in such cases.
if this is the case in development versions as well, we need a ticket to resolve it

BourgoisMickael replied:
Here a ticket to keep track: [CLDSRV-644](https://scality.atlassian.net/browse/CLDSRV-644)

For example with OKMS we did not ask to use debug logs and they ran without it, so having it in info for a while for a few NFI is helpful

anurag4DSB commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,144 @@
+    //     // eslint-disable-next-line no-param-reassign
+    //     objMD.replicationInfo = Object.assign({},
+    //         objMD.replicationInfo, replicationInfo);
+    // }
```
Do we still need this code?
We should not merge commented code.

BourgoisMickael replied:
I did not check yet if the migration (using CopyObject as an operation) impacts backbeat CRR or bucket notifications as outscale don't use it.

Something I planned to check for 9.5

BourgoisMickael replied:
Removed the comment: https://github.com/scality/cloudserver/pull/5788/commits/53a76a671a00ff266374dffbf3f60783382fc089

anurag4DSB commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,144 @@
+    //         objMD.replicationInfo, replicationInfo);
+    // }
+
+    log.info('reformating object encryption key', { oldKey: key, newKey, skipUpdate: opts.skipObjectUpdate });
```
Same for this one, why do we have info here

BourgoisMickael replied:
see previous comment

anurag4DSB commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,144 @@
+
+    const { previousKeyType, previousKeyProtocol, previousKeyProvider } = config.sseMigration;
+    if (!previousKeyType) {
+        // TODO future dynamic evaluation of sse key
```
If not already, please create a ticket and add it here in the comment for reference.

BourgoisMickael replied:
We are waiting for customer's need for other use cases of migration (and it might never happen, if only a few use SSE and none have mixed intern and extern KMS).

For that I think ticket would come from product if it ever happen.
This comment is rather a placeholder, maybe I should rather not write `TODO` inthis

BourgoisMickael replied:
Replaced the todo: https://github.com/scality/cloudserver/pull/5788/commits/53a76a671a00ff266374dffbf3f60783382fc089

anurag4DSB commented on lib/api/apiUtils/bucket/updateEncryption.js:
```diff
@@ -0,0 +1,144 @@
+    const { previousKeyType, previousKeyProtocol, previousKeyProvider } = config.sseMigration;
+    if (!previousKeyType) {
+        // TODO future dynamic evaluation of sse key
+        process.emitWarning('Only explicit KMS encryption key type migration is supported for now', {
+            name: 'S3C-9949',
+            detail: 'You need to set config.sseMigration.previousKeyType to int or ext',
+        });
```
Do we really want to put the ticket number here?

BourgoisMickael replied:
I wrote this warning when I did not completely validated the config.

Now the config is validated (https://github.com/scality/cloudserver/pull/5788/files#diff-b06d72fe5b9eba1ab27b65f49f4fc461894ee5bf7507753734fdb5fff1eee387R1529)

This branch is not reachable anymore I can remove the emitWarning

BourgoisMickael replied:
Dropped the lines: https://github.com/scality/cloudserver/pull/5788/commits/53a76a671a00ff266374dffbf3f60783382fc089

anurag4DSB commented on lib/kms/wrapper.js:
```diff
@@ -31,35 +38,127 @@ function getScalityKms() {
+        client = new KMIPClient(kmipConfig);
+        implName = 'kmip';
+    } else if (kms === 'aws') {
+        const awsConfig = { kmsAWS: config.kmsAWS };
```
Like others do we need a validation here as well

BourgoisMickael replied:
Done: https://github.com/scality/cloudserver/pull/5788/commits/fd1ca6bd2ef29c59a6d4c478c822ffd818aa7e0e

anurag4DSB commented on lib/kms/wrapper.js:
```diff
@@ -31,35 +38,127 @@ function getScalityKms() {
-    const kmipConfig = { kmip: config.kmip };
-    if (!kmipConfig.kmip) {
-        throw new Error('KMIP KMS driver configuration is missing.');
+function getClient(kms) {
```
Not picking.
This function is getting too complex, too many if/else
We could use a strategy here
Something like, may be lowercase part is not needed.
This keeps each backend’s logic self-contained, makes adding new ones as simple as extending the map, and consolidates your error handling and ensures error consistency. I would say standardize on either always throwing for missing config or letting the client constructors fail. At a minimum, document the behavior clearly.
```
const implementations = {
  mem:  { factory: () => inMemory,          name: 'memoryKms' },
  file: { factory: () => file,              name: 'fileKms' },
  cdmi: { factory: () => file,              name: 'fileKms' },
  scality: { factory: () => {
      const { scalityKMS, scalityKMSImpl } = getScalityKms();
      return { client: scalityKMS, implName: scalityKMSImpl };
    }
  },
  kmip: { factory: () => {
      if (!config.kmip) throw new Error('KMIP config missing');
      return { client: new KMIPClient({ kmip: config.kmip }), implName: 'kmip' };
    }
  },
  aws: { factory: () => {
      if (!config.kmsAWS) throw new Error('AWS config missing');
      return { client: new KmsAWSClient({ kmsAWS: config.kmsAWS }), implName: 'aws' };
    }
  },
};

function getClient(kms) {
  kms = (kms || '').toLowerCase();
  const impl = implementations[kms];
  if (!impl) throw new Error(`KMS backend is not configured: ${kms}`);
  return impl.factory();
}
```

BourgoisMickael replied:
Yep this is a very nice way, I've done it here: https://github.com/scality/cloudserver/pull/5788/commits/fd1ca6bd2ef29c59a6d4c478c822ffd818aa7e0e

anurag4DSB commented on lib/api/objectPutPart.js:
```diff
@@ -218,6 +234,7 @@ function objectPutPart(authInfo, request, streamingV4Params, log,
                     return next(err, destinationBucket);
                 }
                 // if data backend handles mpu, skip to end of waterfall
+                // TODO (artesca) data backend should return SSE to include in response headers
```
Can we back this up by a ticket as well.

BourgoisMickael replied:
https://scality.atlassian.net/browse/CLDSRV-640

BourgoisMickael replied:
Added a ref https://github.com/scality/cloudserver/pull/5788/commits/05d6766bd03952b5b3f3ac2c7fc7acef0ab2bceb

anurag4DSB commented on lib/api/apiUtils/object/sseHeaders.js:
```diff
@@ -0,0 +1,18 @@
+const { config } = require('../../../Config');
+const { getKeyIdFromArn } = require('arsenal/build/lib/network/KMSInterface');
+
+function setSSEHeaders(headers, algo, kmsKey) {
+    if (algo) {
+        // eslint-disable-next-line no-param-reassign
+        headers['x-amz-server-side-encryption'] = algo;
+        if (kmsKey && algo === 'aws:kms') {
+            // eslint-disable-next-line no-param-reassign
+            headers['x-amz-server-side-encryption-aws-kms-key-id'] =
+                config.kmsHideScalityArn ? getKeyIdFromArn(kmsKey) : kmsKey;
+        }
+    }
+}
```
General question, instead of mutating, why don't we return a fresh headers?

BourgoisMickael replied:
To follow how adding other headers have been implemented so far. As this is the end of request handling, it is more efficient to use the same object

- https://github.com/scality/cloudserver/blob/0f788abd327750d2a58d630babf1df3cfb423bca/lib/api/apiUtils/object/expirationHeaders.js#L130-L133 This function doesn't return anything it modify input headers
- https://github.com/scality/cloudserver/blob/0f788abd327750d2a58d630babf1df3cfb423bca/lib/api/objectGetACL.js#L194

- https://github.com/scality/cloudserver/blob/0f788abd327750d2a58d630babf1df3cfb423bca/lib/api/objectCopy.js#L604-L624

anurag4DSB commented on package.json:
```diff
@@ -20,7 +20,7 @@
   "dependencies": {
     "@hapi/joi": "^17.1.0",
-    "arsenal": "git+https://github.com/scality/arsenal#7.70.4-2",
+    "arsenal": "git+https://github.com/scality/arsenal#improvement/ARSN-485-kms-provider",
```
Adding comment here to change this before merge

BourgoisMickael replied:
There: https://github.com/scality/cloudserver/pull/5788/commits/8ff7b58ccc01ec3a6360e22730dbafface11eede

anurag4DSB reviewed (approved):
Thanks for the changes

BourgoisMickael committed:
CLDSRV-636: Allow previous KMS for MPU parts

BourgoisMickael committed:
CLDSRV-636: Migrate SSE on CopyObject and init MPU
